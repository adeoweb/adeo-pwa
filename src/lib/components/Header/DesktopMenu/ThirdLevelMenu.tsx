import React, { FunctionComponent } from 'react';
import SimpleError from 'src/lib/components/SimpleError';
import { TCategoryTree } from 'src/lib/types/graphql/Category';
import { Link } from 'src/lib/drivers';

interface IMenuProps {
    menuChildren: TCategoryTree[];
}

const ThirdLevelMenu: FunctionComponent<IMenuProps> = ({ menuChildren }) => {
    if (!Array.isArray(menuChildren)) {
        return <SimpleError />;
    }

    if (!menuChildren) {
        return null;
    }

    return (
        <ul>
            {menuChildren.map(({ name, url_path: urlPath }, index) => {
                const itemKey = `${name}-${index}`;

                if (!name || !urlPath) {
                    return null;
                }

                return (
                    <li key={itemKey}>
                        <Link to={urlPath}>{name}</Link>
                    </li>
                );
            })}
        </ul>
    );
};

export default ThirdLevelMenu;
