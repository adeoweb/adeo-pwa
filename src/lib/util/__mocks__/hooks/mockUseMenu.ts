import { TUseMenu } from 'src/peregrine/lib/talons/adeoweb/Menu/useMenu';

const mock: TUseMenu = {
    error: null,
    children: [
        {
            children_count: '2',
            id: 10,
            include_in_menu: 1,
            name: 'Mock name',
            position: 2,
            url_path: 'mockurl.html',
            children: [
                {
                    children_count: '2',
                    id: 11,
                    include_in_menu: 1,
                    name: 'Mock second name - 1',
                    position: 1,
                    url_path: 'urlpath/path.html',
                    children: [
                        {
                            children_count: '0',
                            id: 123,
                            include_in_menu: 1,
                            name: 'Mock third name - 1',
                            position: 1,
                            url_path: 'urlpath/path.html',
                            children: []
                        },
                        {
                            children_count: '0',
                            id: 124,
                            include_in_menu: 1,
                            name: 'Mock third name - 2',
                            position: 2,
                            url_path: 'urlpath/path.html',
                            children: []
                        }
                    ]
                },
                {
                    children_count: '0',
                    id: 12,
                    include_in_menu: 1,
                    name: 'Mock second name - 1',
                    position: 2,
                    url_path: 'urlpath/path.html',
                    children: []
                }
            ]
        }
    ]
};

export default mock;
