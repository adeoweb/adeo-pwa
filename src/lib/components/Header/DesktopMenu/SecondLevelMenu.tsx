import React, { FunctionComponent } from 'react';
import { Col, Row } from 'react-bootstrap';
import ThirdLevelMenu from 'src/lib/components/Header/DesktopMenu/ThirdLevelMenu';
import SimpleError from 'src/lib/components/SimpleError';
import { Link } from 'src/lib/drivers';
import { TCategoryTree } from 'src/lib/types/graphql/Category';

interface ISecondLevelMenuProps {
    menuChildren: TCategoryTree[];
    isActive: boolean;
}

const SecondLevelMenu: FunctionComponent<ISecondLevelMenuProps> = ({
    menuChildren,
    isActive
}) => {
    const showMenu = isActive ? 'block' : 'none';

    if (!Array.isArray(menuChildren)) {
        return <SimpleError />;
    }

    return (
        <Col className="megamenu" style={{ display: showMenu }}>
            <Row className="row">
                <Col lg={8}>
                    <Row>
                        {menuChildren.map(
                            (
                                {
                                    name,
                                    url_path: urlPath,
                                    children: thirdLevelMenuChildren
                                },
                                index
                            ) => {
                                const itemKey = `${name}-${index}`;

                                if (!name || !urlPath) {
                                    return null;
                                }

                                return (
                                    <Col key={itemKey} lg={4}>
                                        <div className="menu-title">
                                            <Link to={urlPath}>{name}</Link>
                                        </div>
                                        {thirdLevelMenuChildren && (
                                            <ThirdLevelMenu
                                                menuChildren={
                                                    thirdLevelMenuChildren
                                                }
                                            />
                                        )}
                                    </Col>
                                );
                            }
                        )}
                    </Row>
                </Col>
            </Row>
        </Col>
    );
};

export default SecondLevelMenu;
