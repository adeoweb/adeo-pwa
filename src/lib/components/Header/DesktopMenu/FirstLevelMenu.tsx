import React, { FunctionComponent, useState } from 'react';
import { Link } from 'src/lib/drivers';
import SecondLevelMenu from 'src/lib/components/Header/DesktopMenu/SecondLevelMenu';
import SimpleError from 'src/lib/components/SimpleError';
import { TCategoryTree } from 'src/lib/types/graphql/Category';

interface IFirstLevelMenuProps {
    menuChildren: TCategoryTree[];
}

let hoverTimer: any;

const FirstLevelMenu: FunctionComponent<IFirstLevelMenuProps> = ({
    menuChildren
}) => {
    const [active, setActive] = useState(0);
    const hoverDelay: number = 200;

    const toggleMenu = ({ index }: { index: number }) => {
        clearTimeout(hoverTimer);

        hoverTimer = setTimeout(() => {
            setActive(index);
        }, hoverDelay);
    };

    if (!Array.isArray(menuChildren)) {
        return <SimpleError />;
    }

    return (
        <nav className="main-nav">
            <ul
                className="menu sf-arrows sf-js-enabled"
                style={{ touchAction: 'pan-y' }}
            >
                {menuChildren.map(
                    (
                        { name, url_path: urlPath, children: secondLevelMenu },
                        index
                    ) => {
                        const currentIndex = index + 1;
                        const itemKey = `${name}-${index}`;
                        const isActive = active === currentIndex;
                        const haveChilds =
                            Array.isArray(secondLevelMenu) &&
                            secondLevelMenu.length > 0;
                        const showSecondLevel = isActive ? 'show' : '';

                        if (!name || !urlPath) {
                            return;
                        }

                        return (
                            <li
                                key={itemKey}
                                className={`megamenu-container ${showSecondLevel}`}
                                onMouseEnter={() => {
                                    toggleMenu({ index: currentIndex });
                                }}
                                onMouseLeave={() => toggleMenu({ index: 0 })}
                            >
                                <Link
                                    className={haveChilds ? 'sf-with-ul' : ''}
                                    to={urlPath}
                                >
                                    {name}
                                </Link>
                                {haveChilds && secondLevelMenu && (
                                    <SecondLevelMenu
                                        menuChildren={secondLevelMenu}
                                        isActive={isActive}
                                    />
                                )}
                            </li>
                        );
                    }
                )}
            </ul>
        </nav>
    );
};

export default FirstLevelMenu;
