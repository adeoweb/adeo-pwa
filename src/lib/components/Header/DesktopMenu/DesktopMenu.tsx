import React, { FunctionComponent } from 'react';
import { useMenu } from 'src/peregrine/lib/talons/adeoweb/Menu/useMenu';
import GET_MENU_QUERY from 'src/lib/queries/getMenu.graphql';
import FirstLevelMenu from 'src/lib/components/Header/DesktopMenu/FirstLevelMenu';
import SimpleError from 'src/lib/components/SimpleError';

const DesktopMenu: FunctionComponent = () => {
    const { error, children: firstLevelMenu } = useMenu({
        getMenuQuery: GET_MENU_QUERY
    });

    if (error || !Array.isArray(firstLevelMenu)) {
        return <SimpleError />;
    }

    return (
        <nav className="main-nav">
            <FirstLevelMenu menuChildren={firstLevelMenu} />
        </nav>
    );
};

export default DesktopMenu;
