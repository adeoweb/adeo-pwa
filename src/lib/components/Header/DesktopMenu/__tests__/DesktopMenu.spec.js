import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import testRenderer from 'react-test-renderer';

import DesktopMenu from 'src/lib/components/Header/DesktopMenu';
import mockUseMenu from 'src/lib/util/__mocks__/hooks/mockUseMenu';

jest.mock('src/peregrine/lib/talons/adeoweb/Menu/useMenu', () => {
    return {
        useMenu: jest.fn(() => mockUseMenu)
    };
});

test('Renders correctly', () => {
    const component = testRenderer.create(
        <Router>
            <DesktopMenu />
        </Router>
    );
    expect(component.toJSON()).toMatchSnapshot();
});
