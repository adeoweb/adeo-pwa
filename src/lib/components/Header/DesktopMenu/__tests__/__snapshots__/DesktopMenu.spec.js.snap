// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<nav
  className="main-nav"
>
  <nav
    className="main-nav"
  >
    <ul
      className="menu sf-arrows sf-js-enabled"
      style={
        Object {
          "touchAction": "pan-y",
        }
      }
    >
      <li
        className="megamenu-container "
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          className="sf-with-ul"
          href="/mockurl.html"
          onClick={[Function]}
        >
          Mock name
        </a>
        <div
          className="megamenu col"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <div
            className="row row"
          >
            <div
              className="col-lg-8"
            >
              <div
                className="row"
              >
                <div
                  className="col-lg-4"
                >
                  <div
                    className="menu-title"
                  >
                    <a
                      href="/urlpath/path.html"
                      onClick={[Function]}
                    >
                      Mock second name - 1
                    </a>
                  </div>
                  <ul>
                    <li>
                      <a
                        href="/urlpath/path.html"
                        onClick={[Function]}
                      >
                        Mock third name - 1
                      </a>
                    </li>
                    <li>
                      <a
                        href="/urlpath/path.html"
                        onClick={[Function]}
                      >
                        Mock third name - 2
                      </a>
                    </li>
                  </ul>
                </div>
                <div
                  className="col-lg-4"
                >
                  <div
                    className="menu-title"
                  >
                    <a
                      href="/urlpath/path.html"
                      onClick={[Function]}
                    >
                      Mock second name - 1
                    </a>
                  </div>
                  <ul />
                </div>
              </div>
            </div>
          </div>
        </div>
      </li>
    </ul>
  </nav>
</nav>
`;
